### PR TITLE
add hooks for measuring memory usage in a job

### LINKED ISSUE #170: create hook to measure memory usage
PR https://github.com/EESSI/docs/pull/262 introduces a convenient way to measure the memory usage of a reframe test via cgroups

we can take this a step further and make a hook that checks:
- if slurm is used
- if slurm is configured with cgroups:
```
scontrol show config |grep "task/cgroup"
TaskPlugin              = task/cgroup,task/affinity
```
- which version of cgroups is used (v1 or v2)
- location of the cgroup hierarchy:
```
scontrol show config |grep "CgroupMountpoint"
CgroupMountpoint        = /sys/fs/cgroup
```

### PR DESCRIPTION
**Pull request by smoors**
fixes #170

### CODE REVIEW

File: eessi/testsuite/hooks.py [hunk 1]
```diff
@@ -695,3 +698,33 @@ def _check_always_request_gpus(test: rfm.RegressionTest):
     if FEATURES[ALWAYS_REQUEST_GPUS] in test.current_partition.features and not test.num_gpus_per_node:
         test.num_gpus_per_node = test.default_num_gpus_per_node
         log(f'num_gpus_per_node set to {test.num_gpus_per_node} for partition {test.current_partition.name}')
+
+
+def write_memory_usage(test: rfm.RegressionTest):
+    """
+    Write the memory usage into the job output file if we are in a Slurm job and if cgroups is enabled in Slurm
```
Comment (casparvl) [inline eessi/testsuite/hooks.py:705]: Maybe make it even more explicit how this should be called:

```
    @run_after('setup')
    def measure_memory_usage(self):
        """
        Measure memory useage
        """
        hooks.write_memory_usage(self)
```

**Comment by casparvl** [comment]
Works like a charm. Added:

```
    @performance_function('MiB', perf_key='memory')
    def perf_mem(self):
        return hooks.extract_memory(self)

    @run_after('setup')
    def measure_memory_usage(self):
        """
        Measure memory useage
        """
        hooks.write_memory_usage(self)
```

to the TensorFlow test. Rerunning it:

```
{EESSI 2023.06} (eessi_test_venv) [casparl@tcn1 EESSI]$ reframe -c test-suite/eessi/testsuite/tests/apps/tensorflow/tensorflow.py --run -t CI -t "1_2_node|1_node|2_nodes"
[ReFrame Setup]
  version:           4.7.0-dev.3+20ccf7af
  command:           "/gpfs/home4/casparl/EESSI/eessi_test_venv/bin/reframe -c test-suite/eessi/testsuite/tests/apps/tensorflow/tensorflow.py --run -t CI -t '1_2_node|1_node|2_nodes'"
  launched by:       casparl@tcn1.local.snellius.surf.nl
  working directory: '/gpfs/home4/casparl/EESSI'
  settings files:    '<builtin>', '/gpfs/home4/casparl/EESSI/test-suite/config/surf_snellius.py'
  check search path: (R) '/gpfs/home4/casparl/EESSI/test-suite/eessi/testsuite/tests/apps/tensorflow/tensorflow.py'
  stage directory:   '/scratch-shared/casparl/reframe_output/staging'
  output directory:  '/gpfs/home4/casparl/EESSI/reframe_runs/output'
  log files:         '/gpfs/home4/casparl/EESSI/reframe_runs/logs/reframe_20240904_161725.log'

[==========] Running 3 check(s)
[==========] Started on Wed Sep  4 16:17:28 2024+0200

[----------] start processing checks
[ RUN      ] EESSI_TensorFlow %scale=2_nodes %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /cbc475c5 @snellius:rome+default
[ RUN      ] EESSI_TensorFlow %scale=2_nodes %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /cbc475c5 @snellius:genoa+default
[ RUN      ] EESSI_TensorFlow %scale=1_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /9864d0f5 @snellius:rome+default
[ RUN      ] EESSI_TensorFlow %scale=1_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /9864d0f5 @snellius:genoa+default
[ RUN      ] EESSI_TensorFlow %scale=1_2_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /a8654f02 @snellius:rome+default
[ RUN      ] EESSI_TensorFlow %scale=1_2_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /a8654f02 @snellius:genoa+default
[       OK ] (1/6) EESSI_TensorFlow %scale=1_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /9864d0f5 @snellius:rome+default
P: perf: 57417.794296696855 img/s (r:0, l:None, u:None)
P: memory: 3960 MiB (r:0, l:None, u:None)
[       OK ] (2/6) EESSI_TensorFlow %scale=2_nodes %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /cbc475c5 @snellius:rome+default
P: perf: 105571.9017506862 img/s (r:0, l:None, u:None)
P: memory: 3889 MiB (r:0, l:None, u:None)
[       OK ] (3/6) EESSI_TensorFlow %scale=1_2_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /a8654f02 @snellius:rome+default
P: perf: 35363.18039135996 img/s (r:0, l:None, u:None)
P: memory: 2000 MiB (r:0, l:None, u:None)
[       OK ] (4/6) EESSI_TensorFlow %scale=1_2_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /a8654f02 @snellius:genoa+default
P: perf: 59394.48800573312 img/s (r:0, l:None, u:None)
P: memory: 2739 MiB (r:0, l:None, u:None)
[       OK ] (5/6) EESSI_TensorFlow %scale=1_node %module_name=TensorFlow/2.13.0-foss-2023a %device_type=cpu /9864d0f5 @snellius:genoa+default
P: perf: 96415.4007564372 img/s (r:0, l:None, u:None)
P: memory: 4967 MiB (r:0, l:None, u:None)
```

Perfect! (last test seems to hang, but I'm sure that's unrelated)

File: eessi/testsuite/hooks.py (hunk 1 above)
Comment (smoors) [reply eessi/testsuite/hooks.py:705]: good suggestion, added in https://github.com/EESSI/test-suite/pull/174/commits/317a2a8966e36c053613e04b628d616c8ec6b803
i also renamed the hooks to make it a bit more consistent

**Review by casparvl — APPROVED** [review]
Lgtm!

**Pull request merged**